walker.go

```diff
@@ -0,0 +1,429 @@
+		return err
+	}
+
+	w.extendPath(child)
+
+	return w.visitActiveNode(visitor)
+}
+
+// Fetch the child from the `ActiveNode` through the `FetchChild`
+// method of the `NavigableNode` interface.
+func (w *Walker) fetchChild() (NavigableNode, error) {
+	if w.currentDepth == -1 {
+		// First time `down()` is called, `currentDepth` is -1,
+		// return the root node. Don't check available child nodes
+		// (as the `Walker` is not actually on any node just yet
+		// and `ActiveChildIndex` is of no use yet).
+		return w.path[0], nil
+	}
+
+	// Check if the child to fetch exists.
+	if w.ActiveChildIndex() >= w.ActiveNode().ChildTotal() {
+		return nil, ErrDownNoChild
+	}
+
+	return w.ActiveNode().FetchChild(w.ctx, w.ActiveChildIndex())
+
+	// TODO: Maybe call `extendPath` here and hide it away
+	// from `down`.
+}
+
+// Increase the `currentDepth` and extend the `path` to the fetched
+// `child` node (which now becomes the new `ActiveNode`)
+func (w *Walker) extendPath(child NavigableNode) {
+	w.currentDepth++
+
+	// Extend the slices if needed (doubling its capacity).
+	if w.currentDepth >= len(w.path) {
+		w.path = append(w.path, make([]NavigableNode, len(w.path))...)
+		w.childIndex = append(w.childIndex, make([]uint, len(w.childIndex))...)
+		// TODO: Check the performance of this grow mechanism.
+	}
+
+	// `child` now becomes the `ActiveNode()`.
+	w.path[w.currentDepth] = child
+	w.childIndex[w.currentDepth] = 0
+}
+
+// Call the `Visitor` on the `ActiveNode`. This function should only be
+// called from `down`. This is a wrapper function to `Visitor` to process
+// the `Pause` signal and do other minor checks (taking this logic away
+// from `down`).
+func (w *Walker) visitActiveNode(visitor Visitor) error {
+	if visitor == nil {
+		return nil
+		// No need to check `pauseRequested` as `Pause` should
+		// only be called from within the `Visitor`.
+	}
+
+	err := visitor(w.ActiveNode())
+
+	if w.pauseRequested {
+		// If a pause was requested make sure an error is returned
+		// that will cause the current walk operation to return. If
+		// `Visitor` didn't return an error set an artificial one
+		// generated by the `Walker`.
+		if err == nil {
+			err = errPauseWalkOperation
+		}
+
+		w.pauseRequested = false
+	}
+
+	return err
+}
+
+// Go up from the `ActiveNode`. The only possible error this method
+// can return is to signal it's already at the root and can't go up.
+func (w *Walker) up() error {
+	if w.currentDepth < 1 {
+		return errUpOnRoot
+	}
+
+	w.currentDepth--
+
+	// w.ActiveNode().Cleanup()
+	// If `Cleanup` is supported this would be the place to call it.
+
+	return nil
+}
+
+// Next increases the child index of the `ActiveNode` to point to the
+// next child (which may exist or may be the end of the available
+// child nodes).
+//
+// This method doesn't change the `ActiveNode`, it just changes where
+// is it pointing to next, it could be interpreted as "turn to the next
+// child".
+func (w *Walker) Next() error {
```

Comment:
Renaming this to `NextChild()` or `SkipChild()` might be clearer — if I’m calling this from inside an `Iterate()` visitor, “next” might sound more like it’s just getting the next node that would be iterated.